sftp services always failed on python binding

```shell
tests/test_read.py:30: Error
_______________________________ test_async_read ________________________________

service_name = 'sftp', operator = Operator("sftp", root="/upload/sftp_test/")
async_operator = AsyncOperator("sftp", root="/upload/sftp_test/")

    @pytest.mark.asyncio
    @pytest.mark.need_capability("read", "write", "delete")
    async def test_async_read(service_name, operator, async_operator):
        size = randint(1, 1024)
        filename = f"random_file_{str(uuid4())}"
        content = os.urandom(size)
>       await async_operator.write(filename, content)
E       opendal.Error: Unexpected (permanent) at write, context: { service: sftp, path: random_file_a3a20165-da47-4689-b85e-f00b2bf2f453 } => ssh error, source: failed to connect to the remote host
```

--- a/.github/scripts/behavior_test/plan.py
+++ b/.github/scripts/behavior_test/plan.py
@@ -214,11 +214,6 @@ def generate_binding_python_cases(
 ) -> list[dict[str, str]]:
     cases = unique_cases(cases)
 
-    # REMOVE ME after https://github.com/apache/incubator-opendal/issues/3429 addressed.
-    #
-    # Sftp can't pass on python bindings, remove it for now.
-    cases = [v for v in cases if v["service"] != "sftp"]
-
     if os.getenv("GITHUB_IS_PUSH") == "true":
         return cases
 

--- a/.github/services/azblob/azurite_azblob/action.yml
+++ b/.github/services/azblob/azurite_azblob/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup azurite azblob service
       shell: bash
       working-directory: fixtures/azblob
-      run: docker-compose -f docker-compose-azurite.yml up -d
+      run: docker compose -f docker-compose-azurite.yml up -d --wait --wait-timeout 20
     - name: Setup test bucket
       shell: bash
       run: |

--- a/.github/services/etcd/etcd-cluster/action.yml
+++ b/.github/services/etcd/etcd-cluster/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup etcd cluster
       shell: bash
       working-directory: fixtures/etcd
-      run: docker-compose -f docker-compose-cluster.yml up -d
+      run: docker compose -f docker-compose-cluster.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/etcd/etcd-tls/action.yml
+++ b/.github/services/etcd/etcd-tls/action.yml
@@ -30,7 +30,7 @@ runs:
     - name: Setup etcd tls
       shell: bash
       working-directory: fixtures/etcd
-      run: docker-compose -f docker-compose-standalone-tls.yml up -d
+      run: docker compose -f docker-compose-standalone-tls.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/etcd/etcd/action.yml
+++ b/.github/services/etcd/etcd/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup etcd cluster
       shell: bash
       working-directory: fixtures/etcd
-      run: docker-compose -f docker-compose-standalone.yml up -d
+      run: docker compose -f docker-compose-standalone.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/http/caddy/action.yml
+++ b/.github/services/http/caddy/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup http in caddy
       shell: bash
       working-directory: fixtures/http
-      run: docker-compose -f docker-compose-caddy.yml up -d
+      run: docker compose -f docker-compose-caddy.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/http/nginx/action.yml
+++ b/.github/services/http/nginx/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup http in nginx
       shell: bash
       working-directory: fixtures/http
-      run: docker-compose -f docker-compose-nginx.yml up -d
+      run: docker compose -f docker-compose-nginx.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/memcached/memcached/action.yml
+++ b/.github/services/memcached/memcached/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup memcached server
       shell: bash
       working-directory: fixtures/memcached
-      run: docker-compose -f docker-compose-memcached.yml up -d
+      run: docker compose -f docker-compose-memcached.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/mongodb/mongodb_with_basic_auth/action.yml
+++ b/.github/services/mongodb/mongodb_with_basic_auth/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup MongoDB Server
       shell: bash
       working-directory: fixtures/mongodb
-      run: docker-compose -f docker-compose-basic-auth.yml up -d
+      run: docker compose -f docker-compose-basic-auth.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/mongodb/mongodb_with_no_auth/action.yml
+++ b/.github/services/mongodb/mongodb_with_no_auth/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup MongoDB Server
       shell: bash
       working-directory: fixtures/mongodb
-      run: docker-compose -f docker-compose-no-auth.yml up -d
+      run: docker compose -f docker-compose-no-auth.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/mysql/mysql/action.yml
+++ b/.github/services/mysql/mysql/action.yml
@@ -26,7 +26,7 @@ runs:
       working-directory: fixtures/mysql
       run: |
         apt update && apt install -y mysql-client
-        docker-compose -f docker-compose.yml up -d
+        docker compose -f docker-compose.yml up -d --wait --wait-timeout 20
         while ! mysql -h localhost -P 3306 --protocol=tcp -u root -p'password' -e "SELECT 1"; do
             echo "Waiting for MySQL..."
             sleep 1

--- a/.github/services/postgresql/postgresql/action.yml
+++ b/.github/services/postgresql/postgresql/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup PostgreSQL Server
       shell: bash
       working-directory: fixtures/postgresql
-      run: docker-compose -f docker-compose.yml up -d
+      run: docker compose -f docker-compose.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/redis/dragonfly/action.yml
+++ b/.github/services/redis/dragonfly/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup Dragonfly Server
       shell: bash
       working-directory: fixtures/redis
-      run: docker-compose -f docker-compose-dragonfly.yml up -d
+      run: docker compose -f docker-compose-dragonfly.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/redis/kvrocks/action.yml
+++ b/.github/services/redis/kvrocks/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup Kvrocks Server
       shell: bash
       working-directory: fixtures/redis
-      run: docker-compose -f docker-compose-kvrocks.yml up -d
+      run: docker compose -f docker-compose-kvrocks.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/redis/redis/action.yml
+++ b/.github/services/redis/redis/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup Redis Server
       shell: bash
       working-directory: fixtures/redis
-      run: docker-compose -f docker-compose-redis.yml up -d
+      run: docker compose -f docker-compose-redis.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/redis/redis_with_cluster/action.yml
+++ b/.github/services/redis/redis_with_cluster/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup Redis Cluster
       shell: bash
       working-directory: fixtures/redis
-      run: docker-compose -f docker-compose-redis-cluster.yml up -d
+      run: docker compose -f docker-compose-redis-cluster.yml up -d --wait --wait-timeout 20
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/s3/minio_s3/action.yml
+++ b/.github/services/s3/minio_s3/action.yml
@@ -24,7 +24,7 @@ runs:
     - name: Setup MinIO Server
       shell: bash
       working-directory: fixtures/s3
-      run: docker-compose -f docker-compose-minio.yml up -d
+      run: docker compose -f docker-compose-minio.yml up -d --wait --wait-timeout 20
     - name: Setup test bucket
       shell: bash
       env:

--- a/.github/services/s3/minio_s3_with_anonymous/action.yml
+++ b/.github/services/s3/minio_s3_with_anonymous/action.yml
@@ -25,7 +25,7 @@ runs:
       shell: bash
       working-directory: fixtures/s3
       run: |
-        docker-compose -f docker-compose-minio.yml up -d
+        docker compose -f docker-compose-minio.yml up -d --wait --wait-timeout 20
     - name: Setup test bucket
       shell: bash
       env:

--- a/.github/services/sftp/sftp/action.yml
+++ b/.github/services/sftp/sftp/action.yml
@@ -26,26 +26,6 @@ runs:
       working-directory: fixtures/sftp
       run: |
         docker compose -f docker-compose-sftp.yml up -d --wait --wait-timeout 20
-        set +ex
-        username="foo"
-        server="127.0.0.1"
-        port=2222
-        know_hosts_strategy="no"
-        identity_file="${{ github.workspace }}/target/ssh/id_rsa"
-        while :; do
-            (
-              sftp -oPort=$port -o StrictHostKeyChecking=$know_hosts_strategy -o IdentityFile=$identity_file $username@$server << EOF
-              bye
-        EOF
-            ) 2>/dev/null
-            if [[ $? -eq 0 ]]; then
-                echo "SFTP is available, proceeding..."
-                break
-            else
-                echo "Waiting for SFTP to be available..."
-                sleep 1
-            fi
-        done
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/sftp/sftp/action.yml
+++ b/.github/services/sftp/sftp/action.yml
@@ -25,7 +25,7 @@ runs:
       shell: bash
       working-directory: fixtures/sftp
       run: |
-        docker-compose -f docker-compose-sftp.yml up -d
+        docker compose -f docker-compose-sftp.yml up -d --wait --wait-timeout 20
         set +ex
         username="foo"
         server="127.0.0.1"

--- a/.github/services/sftp/sftp/action.yml
+++ b/.github/services/sftp/sftp/action.yml
@@ -24,7 +24,28 @@ runs:
     - name: Setup Sftp
       shell: bash
       working-directory: fixtures/sftp
-      run: docker-compose -f docker-compose-sftp.yml up -d
+      run: |
+        docker-compose -f docker-compose-sftp.yml up -d
+        set +ex
+        username="foo"
+        server="127.0.0.1"
+        port=2222
+        know_hosts_strategy="no"
+        identity_file="${{ github.workspace }}/target/ssh/id_rsa"
+        while :; do
+            (
+              sftp -oPort=$port -o StrictHostKeyChecking=$know_hosts_strategy -o IdentityFile=$identity_file $username@$server << EOF
+              bye
+        EOF
+            ) 2>/dev/null
+            if [[ $? -eq 0 ]]; then
+                echo "SFTP is available, proceeding..."
+                break
+            else
+                echo "Waiting for SFTP to be available..."
+                sleep 1
+            fi
+        done
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/sftp/sftp_with_default_root/action.yml
+++ b/.github/services/sftp/sftp_with_default_root/action.yml
@@ -26,26 +26,6 @@ runs:
       working-directory: fixtures/sftp
       run: |
         docker compose -f docker-compose-sftp-with-default-root.yml up -d --wait --wait-timeout 20
-        set +ex
-        username="foo"
-        server="127.0.0.1"
-        port=2222
-        know_hosts_strategy="no"
-        identity_file="${{ github.workspace }}/target/ssh/id_rsa"
-        while :; do
-            (
-              sftp -oPort=$port -o StrictHostKeyChecking=$know_hosts_strategy -o IdentityFile=$identity_file $username@$server << EOF
-              bye
-        EOF
-            ) 2>/dev/null
-            if [[ $? -eq 0 ]]; then
-                echo "SFTP is available, proceeding..."
-                break
-            else
-                echo "Waiting for SFTP to be available..."
-                sleep 1
-            fi
-        done
     - name: Setup
       shell: bash
       run: |

--- a/.github/services/sftp/sftp_with_default_root/action.yml
+++ b/.github/services/sftp/sftp_with_default_root/action.yml
@@ -25,7 +25,7 @@ runs:
       shell: bash
       working-directory: fixtures/sftp
       run: |
-        docker-compose -f docker-compose-sftp-with-default-root.yml up -d
+        docker compose -f docker-compose-sftp-with-default-root.yml up -d --wait --wait-timeout 20
         set +ex
         username="foo"
         server="127.0.0.1"

--- a/.github/services/sftp/sftp_with_default_root/action.yml
+++ b/.github/services/sftp/sftp_with_default_root/action.yml
@@ -24,7 +24,28 @@ runs:
     - name: Setup Sftp
       shell: bash
       working-directory: fixtures/sftp
-      run: docker-compose -f docker-compose-sftp-with-default-root.yml up -d
+      run: |
+        docker-compose -f docker-compose-sftp-with-default-root.yml up -d
+        set +ex
+        username="foo"
+        server="127.0.0.1"
+        port=2222
+        know_hosts_strategy="no"
+        identity_file="${{ github.workspace }}/target/ssh/id_rsa"
+        while :; do
+            (
+              sftp -oPort=$port -o StrictHostKeyChecking=$know_hosts_strategy -o IdentityFile=$identity_file $username@$server << EOF
+              bye
+        EOF
+            ) 2>/dev/null
+            if [[ $? -eq 0 ]]; then
+                echo "SFTP is available, proceeding..."
+                break
+            else
+                echo "Waiting for SFTP to be available..."
+                sleep 1
+            fi
+        done
     - name: Setup
       shell: bash
       run: |

--- a/fixtures/sftp/docker-compose-sftp-with-default-root.yml
+++ b/fixtures/sftp/docker-compose-sftp-with-default-root.yml
@@ -19,14 +19,27 @@ version: '3.8'
 
 services:
   sftp-with-default-root:
-    image: atmoz/sftp
+    image: atmoz/sftp:debian
     ports:
       - "2222:22"
     volumes:
       - ./change_root_dir.sh:/etc/sftp.d/change_root_dir
       - ./test_ssh_key.pub:/home/foo/.ssh/keys/id_rsa.pub:ro
+      - ./test_ssh_key:/home/foo/.ssh/keys/id_rsa:ro
+      - ./health-check.sh:/health-check.sh
     ulimits:
       nofile:
         soft: 65536
         hard: 65536
     command: foo::::upload
+    healthcheck:
+      test:
+        [
+          "CMD-SHELL",
+          "bash",
+          "-c",
+          "/health-check.sh"
+        ]
+      interval: 10s
+      timeout: 10s
+      retries: 5

--- a/fixtures/sftp/docker-compose-sftp.yml
+++ b/fixtures/sftp/docker-compose-sftp.yml
@@ -19,13 +19,26 @@ version: '3.8'
 
 services:
   sftp:
-    image: atmoz/sftp
+    image: atmoz/sftp:debian
     ports:
       - "2222:22"
     volumes:
       - ./test_ssh_key.pub:/home/foo/.ssh/keys/id_rsa.pub:ro
+      - ./test_ssh_key:/home/foo/.ssh/keys/id_rsa:ro
+      - ./health-check.sh:/health-check.sh
     ulimits:
       nofile:
         soft: 65536
         hard: 65536
     command: foo::::upload
+    healthcheck:
+      test:
+        [
+          "CMD-SHELL",
+          "bash",
+          "-c",
+          "/health-check.sh"
+        ]
+      interval: 10s
+      timeout: 10s
+      retries: 5

--- a/fixtures/sftp/health-check.sh
+++ b/fixtures/sftp/health-check.sh
@@ -1,0 +1,22 @@
+# !/bin/bash
+
+set +ex
+username="foo"
+server="127.0.0.1"
+port=22
+know_hosts_strategy="no"
+identity_file="/home/foo/.ssh/keys/id_rsa"
+while :; do
+    (
+        sftp -oPort=$port -o StrictHostKeyChecking=$know_hosts_strategy -o IdentityFile=$identity_file $username@$server << EOF
+        bye
+EOF
+    ) 2>/dev/null
+    if [[ $? -eq 0 ]]; then
+        echo "SFTP is available, proceeding..."
+        break
+    else
+        echo "Waiting for SFTP to be available..."
+        sleep 1
+    fi
+done

